provider/qsd.py

```diff
@@ -0,0 +1,473 @@
+
+        LOG_JOB.info(self.qsd_version)
+        self.caps = Capabilities()
+        self._probe_capabilities()
+        self.images = {}
+        self.daemonize = False
+        self.pidfile = None
+        self.pid = None
+
+    def _remove_images(self):
+        for img in self.images.values():
+            if img["image_object"]:
+                LOG_JOB.info("QSD ready to remove image:%s", img["name"])
+                if self.qsd_params.get("remove_image_%s" % img["name"],
+                                       "yes") != "no":
+                    img["image_object"].remove()
+
+    def _fulfil_image_props(self, name, params):
+        """Fulfil image property and prepare image file """
+        img = copy.deepcopy(QsdDaemonDev.raw_image_data)
+        img["name"] = name
+        img["protocol"]["node-name"] = "prot_" + name
+        filename = get_image_filename(params, data_dir.get_data_dir())
+        img["protocol"]["filename"] = filename
+        img["format"]["node-name"] = "fmt_" + name
+        img["format"]["file"] = img["protocol"]["node-name"]
+        img["filter"]["node-name"] = "flt_" + name
+        img["filter"]["file"] = img["format"]["node-name"]
+
+        img["protocol"].update(
+            json.loads(params.get("qsd_image_protocol", "{}")))
+        img["format"].update(json.loads(params.get("qsd_image_format", "{}")))
+        img["filter"].update(json.loads(params.get("qsd_image_filter", "{}")))
+
+        img["export"]["id"] = "id_" + name
+        img["export"]["node-name"] = img["filter"]["node-name"] if \
+            img["filter"]["driver"] else img["format"]["node-name"]
+        img["export"].update(json.loads(params.get("qsd_image_export", "{}")))
+
+        if img["export"]["type"] == "nbd":
+            # The name is necessary. empty value to simply setting in nbd client
+            img["export"]["name"] = ""
+            addr = json.loads(
+                params.get("qsd_image_export_nbd", '{"type":"unix"'))
+
+            img[addr["type"]].update(addr)
+            if addr["type"] == "unix":
+                if not img[addr["type"]]["path"]:
+                    img[addr["type"]]["path"] = "%s/%s_nbd_%s.sock" % (
+                        self.basedir, self.name, name)
+            img["nbd-server"]["addr"].update(img[addr["type"]])
+
+        elif img["export"]["type"] == "vhost-user-blk":
+            img["unix"]["path"] = "%s/%s_vhost_user_%s.sock" % (
+                self.basedir, self.name, name)
+            img["export"]["addr"] = img["unix"]
+        else:
+            raise QsdError("Unknown export type %s " % img["export"]["type"])
+
+        # Prepare image
+        image_created = False
+
+        if name in params.get("images").split():
+            if params.get("force_create_image") == "yes":
+                image_created = True
+            elif params.get(
+                    "create_image") == "yes" and not storage.file_exists(
```

Comment:
when QSD inits, the preprocess of VT should have been finished, right?
Then, storage.file_exists should always returns True here

Reply:
updated